File: src/gaia_lsp/diagnostics.py
"""Diagnostic data model for gaia-lsp."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any


@dataclass(frozen=True)
class Diagnostic:
    """A source diagnostic with 1-based display positions."""

    code: str
    message: str
    severity: str
    line: int
    column: int
    end_line: int | None = None
    end_column: int | None = None
    source: str = "gaia-lsp"
    file: str | None = None

    @property
    def blocking(self) -> bool:
        return self.severity == "error"

    def to_json(self) -> dict[str, Any]:
        end_line = self.end_line if self.end_line is not None else self.line
        end_column = self.end_column if self.end_column is not None else self.column + 1
        payload: dict[str, Any] = {
            "code": self.code,
            "message": self.message,
            "severity": self.severity,
            "source": self.source,
            "blocking": self.blocking,
            "range": {
                "start": {"line": max(self.line - 1, 0), "character": max(self.column - 1, 0)},
                "end": {
                    "line": max(end_line - 1, 0),
                    "character": max(end_column - 1, 0),
                },
            },
            "line": self.line,
            "column": self.column,
        }
        if self.file:
            payload["file"] = self.file
        return payload


def diagnostic(
    code: str,
    message: str,
    severity: str,
    node: Any,
    *,
    path: Path | None = None,
) -> Diagnostic:
    """Create a diagnostic from an AST node-like object."""

    line = int(getattr(node, "lineno", 1) or 1)
    column = int(getattr(node, "col_offset", 0) or 0) + 1
    end_line = int(getattr(node, "end_lineno", line) or line)
    end_column = int(getattr(node, "end_col_offset", column) or column) + 1
    return Diagnostic(
        code=code,
        message=message,
        severity=severity,
        line=line,
        column=column,
        end_line=end_line,
        end_column=end_column,
        file=str(path) if path else None,
    )

File: gaia-vscode/src/diagnostics.ts
import * as path from "node:path";

export interface GaiaPosition {
  line: number;
  character: number;
}

export interface GaiaRange {
  start: GaiaPosition;
  end: GaiaPosition;
}

export interface GaiaDiagnostic {
  code: string;
  message: string;
  severity: string;
  source?: string;
  blocking?: boolean;
  file?: string;
  range?: GaiaRange;
  line?: number;
  column?: number;
}

export interface GaiaCheckResult {
  software: string;
  operation: string;
  path: string;
  ok: boolean;
  blockingDiagnosticCount: number;
  diagnosticCount: number;
  diagnostics: GaiaDiagnostic[];
}

export type DiagnosticSeverityName = "Error" | "Warning" | "Information" | "Hint";

export interface DiagnosticTransfer {
  filePath: string;
  code: string;
  message: string;
  severity: DiagnosticSeverityName;
  source: string;
  range: GaiaRange;
  blocking: boolean;
}

export function parseGaiaCheckOutput(stdout: string): GaiaCheckResult {
  const payload = JSON.parse(stdout) as Partial<GaiaCheckResult>;
  if (payload.software !== "gaia" || payload.operation !== "check") {
    throw new Error("gaia-lsp-tool returned a non-check payload");
  }
  if (!Array.isArray(payload.diagnostics)) {
    throw new Error("gaia-lsp-tool payload is missing diagnostics");
  }
  return {
    software: payload.software,
    operation: payload.operation,
    path: String(payload.path ?? ""),
    ok: Boolean(payload.ok),
    blockingDiagnosticCount: Number(payload.blockingDiagnosticCount ?? 0),
    diagnosticCount: Number(payload.diagnosticCount ?? payload.diagnostics.length),
    diagnostics: payload.diagnostics
  };
}

export function severityName(severity: string, blocking = false): DiagnosticSeverityName {
  const normalized = severity.toLowerCase();
  if (normalized === "error" || blocking) {
    return "Error";
  }
  if (normalized === "warning" || normalized === "warn") {
    return "Warning";
  }
  if (normalized === "info" || normalized === "information") {
    return "Information";
  }
  return "Hint";
}

export function normalizeRange(diagnostic: GaiaDiagnostic): GaiaRange {
  if (diagnostic.range) {
    return clampRange(diagnostic.range);
  }
  const line = Math.max(Number(diagnostic.line ?? 1) - 1, 0);
  const character = Math.max(Number(diagnostic.column ?? 1) - 1, 0);
  return {
    start: { line, character },
    end: { line, character: character + 1 }
  };
}

export function resolveDiagnosticFile(diagnostic: GaiaDiagnostic, fallbackPath: string): string {
  const candidate = diagnostic.file || fallbackPath;
  if (!candidate) {
    return fallbackPath;
  }
  if (path.isAbsolute(candidate)) {
    return path.normalize(candidate);
  }
  const basePath = fallbackPath && path.extname(fallbackPath) ? path.dirname(fallbackPath) : fallbackPath;
  return path.normalize(path.resolve(basePath || process.cwd(), candidate));
}

export function groupDiagnosticsByFile(
  payload: GaiaCheckResult,
  fallbackPath = payload.path
): Record<string, DiagnosticTransfer[]> {
  const grouped: Record<string, DiagnosticTransfer[]> = {};
  for (const diagnostic of payload.diagnostics) {
    const filePath = resolveDiagnosticFile(diagnostic, fallbackPath);
    const item: DiagnosticTransfer = {
      filePath,
      code: diagnostic.code,
      message: diagnostic.message,
      severity: severityName(diagnostic.severity, diagnostic.blocking),
      source: diagnostic.source || "gaia-lsp",
      range: normalizeRange(diagnostic),
      blocking: Boolean(diagnostic.blocking)
    };
    grouped[filePath] = grouped[filePath] || [];
    grouped[filePath].push(item);
  }
  return grouped;
}

export function summarizeCheck(payload: GaiaCheckResult): string {
  const blocking = payload.blockingDiagnosticCount;
  const total = payload.diagnosticCount;
  if (total === 0) {
    return "Gaia LSP: no diagnostics";
  }
  if (blocking > 0) {
    return `Gaia LSP: ${blocking} blocking / ${total} total`;
  }
  return `Gaia LSP: ${total} non-blocking diagnostics`;
}

function clampRange(range: GaiaRange): GaiaRange {
  const start = {
    line: Math.max(Number(range.start.line), 0),
    character: Math.max(Number(range.start.character), 0)
  };
  const end = {
    line: Math.max(Number(range.end.line), start.line),
    character: Math.max(Number(range.end.character), 0)
  };
  if (end.line === start.line && end.character <= start.character) {
    end.character = start.character + 1;
  }
  return { start, end };
}
